prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs

```diff
@@ -514,16 +514,106 @@ softline' = group line'
 hardline :: Doc ann
 hardline = Line
 
+group :: Doc ann -> Doc ann
+group = \doc -> case doc of
+    FlatAlt x y     -> case changesOnFlattening y of
+        HasLine  -> x
+        NoChange -> Union y x
+        Flat y'  -> Union y' x
+    
+    x@(Cat a b)  -> case (changesOnFlattening a, changesOnFlattening b) of
+        (HasLine  , _       ) -> x
+        (_        , HasLine ) -> x
+        (NoChange , NoChange) -> x
+        (NoChange , Flat b' ) -> Cat a (Union b' b)
+        (Flat a'  , NoChange) -> Cat (Union a' a) b
+        (Flat a'  , Flat b' ) -> Union (Cat a' b') (Cat a b)
+
+    Annotated ann x -> Annotated ann (group x)
+    Nest i x        -> Nest i (group x)
+
+    Column f        -> Column (group . f)
+    Nesting f       -> Nesting (group . f)
+    WithPageWidth f -> WithPageWidth (group . f)
+
+    x@Union{}       -> x
+    x@Char{}        -> x
+    x@Text{}        -> x
+    x@Line          -> x
+    x@Empty         -> x
+    -- Should never happen on a valid document
+    x@Fail          -> x
+
+changesOnFlattening :: Doc ann -> FlatteningResult (Doc ann)
+changesOnFlattening = \doc -> case doc of
+    FlatAlt _ y     -> case changesOnFlattening y of
+        HasLine  -> HasLine
+        NoChange -> Flat y
+        Flat y'  -> Flat y'
+    
+    Union x _       -> Flat x
+
+    Cat a b    -> case (changesOnFlattening a, changesOnFlattening b) of
+        (HasLine  , _       ) -> HasLine
+        (_        , HasLine ) -> HasLine
+        (NoChange , NoChange) -> NoChange
+        (Flat a'  , NoChange) -> Flat (Cat a' b)
+        (NoChange , Flat b' ) -> Flat (Cat a b')
+        (Flat a'  , Flat b' ) -> Flat (Cat a' b')
+    
+    Annotated ann x -> Annotated ann <$> (changesOnFlattening x)
+    Nest i x        -> Nest i <$> (changesOnFlattening x)
+
+    Column f        -> Flat (Column (flatten . f))
+    Nesting f       -> Flat (Nesting (flatten . f))
+    WithPageWidth f -> Flat (WithPageWidth (flatten . f))
```

Comment:
Because `changesOnFlattening` is expected to return a value that tells the caller if the doc changed (either by flattening, by encountering a Line or not changing). And you cannot get that information out of a function :/

Reply:
Hm, I don't understand what you're saying here. Both `flatten` and `group` return `Doc`s. And `group` can actually add information by wrapping things in `Union`.

Reply:
Thats what you are thinking about! Yes definitly that is a good idea, I'll test that

Reply:
This fails quickcheck with a huge counterexample again and my own tests with a small one. Now to figure out whats wrong...

Reply:
Counterexample: `column (\_ -> hardline)`. This flattens to `column (\_ -> Fail)` but when using `group` to `column (\_ -> Line)`.

Reply:
Ah, that's very interesting! Would that explain the [test failures](https://github.com/quchen/prettyprinter/pull/126#issuecomment-577439337) I got for the first version of this PR?

Reply:
Yes, or at least the problem was very similar iirc.